rust/operator-binary/src/spark_k8s_controller.rs

```diff
@@ -368,62 +506,160 @@ fn pod_template_config_map(
+    } else {
+        spark_application.submit_job_config_map_name()
+    };
+
+    let mut args = vec![job_commands.join(" ")];
+    if job_config.logging.enable_vector_agent {
+        // Wait for Vector to gather the logs.
+        args.push("sleep 10".into());
+        args.push(shutdown_vector_command(VOLUME_MOUNT_PATH_LOG));
+    }
+
     cb.image(spark_image)
-        .command(vec!["/bin/sh".to_string()])
+        .command(vec!["/bin/bash".to_string(), "-c".to_string()])
+        .args(vec![args.join(" && ")])
         .resources(job_config.resources.into())
-        .args(vec!["-c".to_string(), job_commands.join(" ")])
         .add_volume_mounts(spark_application.spark_job_volume_mounts(s3conn, s3logdir))
         .add_env_vars(env.to_vec())
+        .add_env_var(
+            "SPARK_SUBMIT_OPTS",
+            format!(
+                "-cp /stackable/spark/extra-jars/*:/stackable/spark/jars/* \
+                -Dlog4j.configurationFile={VOLUME_MOUNT_PATH_LOG_CONFIG}/{LOG4J2_CONFIG_FILE}"
+            ),
+        )
         // TODO: move this to the image
-        .add_env_vars(vec![EnvVar {
-            name: "SPARK_CONF_DIR".to_string(),
-            value: Some("/stackable/spark/conf".to_string()),
-            value_from: None,
-        }]);
+        .add_env_var("SPARK_CONF_DIR", "/stackable/spark/conf");
 
     if let Some(image_pull_policy) = spark_application.spark_image_pull_policy() {
         cb.image_pull_policy(image_pull_policy.to_string());
     }
 
-    let mut volumes = vec![Volume {
-        name: String::from(VOLUME_MOUNT_NAME_POD_TEMPLATES),
-        config_map: Some(ConfigMapVolumeSource {
-            name: Some(spark_application.pod_template_config_map_name()),
-            ..ConfigMapVolumeSource::default()
-        }),
-        ..Volume::default()
-    }];
-    volumes.extend(spark_application.volumes(s3conn, s3logdir));
+    let mut volumes = vec![
+        VolumeBuilder::new(VOLUME_MOUNT_NAME_CONFIG)
+            .with_config_map(spark_application.submit_job_config_map_name())
+            .build(),
+        VolumeBuilder::new(VOLUME_MOUNT_NAME_DRIVER_POD_TEMPLATES)
+            .with_config_map(
+                spark_application.pod_template_config_map_name(SparkApplicationRole::Driver),
+            )
+            .build(),
+        VolumeBuilder::new(VOLUME_MOUNT_NAME_EXECUTOR_POD_TEMPLATES)
+            .with_config_map(
+                spark_application.pod_template_config_map_name(SparkApplicationRole::Executor),
+            )
+            .build(),
+    ];
+    volumes.extend(spark_application.volumes(s3conn, s3logdir, &log_config_map));
+
+    let mut containers = vec![cb.build()];
+
+    if job_config.logging.enable_vector_agent {
+        containers.push(vector_container(
+            &ResolvedProductImage {
+                product_version: "".into(),
+                app_version_label: "".into(),
+                image: spark_application.spec.spark_image.clone().unwrap(),
```

Comment:
This can fail and kill the operator?

Reply:
Changed to

```rust
                image: spark_application
                    .spec
                    .spark_image
                    .clone()
                    .context(ObjectHasNoSparkImageSnafu)?,
```

in 9e8ed3f0b7524dc78febf2ca7edbd0c689306684